presto-main/src/main/java/io/prestosql/execution/warnings/statswarnings/JoinWarningsGenerator.java

```diff
@@ -0,0 +1,288 @@
+            PlanNode rightNode = node.getSource();
+            String joinLabel = "SemiJoin";
+            WarningSource left = leftNode.accept(this, context);
+            WarningSource right = rightNode.accept(this, context);
+            generateJoinWarnings(left, right, leftNode, rightNode, joinLabel, id);
+            return combineWarningSources(left, right);
+        }
+
+        @Override
+        public WarningSource visitRemoteSource(RemoteSourceNode node, Void context)
+        {
+            WarningSource warningSource = new WarningSource();
+            warningSource.addStageId(node.getSourceFragmentIds().stream().map(PlanFragmentId::toString).collect(Collectors.toList()));
+            return warningSource;
+        }
+
+        @Override
+        public WarningSource visitTableScan(TableScanNode node, Void context)
+        {
+            WarningSource warningSource = new WarningSource();
+            warningSource.addTableNames(Arrays.asList(node.getTable().toString()));
+            return warningSource;
+        }
+
+        private long getOperatorstats(PlanNode planNode)
+        {
+            List<OperatorStats> operatorStats = operatorStatsList;
+            for (OperatorStats operatorStat : operatorStats) {
+                if (operatorStat.getPlanNodeId().equals(planNode.getId())) {
+                    return operatorStat.getOutputPositions();
+                }
+            }
+            return 0;
+        }
+
+        private void generateJoinWarnings(WarningSource left, WarningSource right, PlanNode leftNode, PlanNode rightNode, String joinLabel, String stageId)
+        {
+            long leftNumberOfRow = getOperatorstats(leftNode);
+            long rightNumberOfRow = getOperatorstats(rightNode);
+
+            if (!(leftNumberOfRow < rightNumberOfRow)) {
+                return;
+            }
+
+            List<String> leftMessageBuilder = new ArrayList();
+            List<String> rightMessageBuilder = new ArrayList();
+            String leftMessage;
+            String rightMessage;
+
+            if (!left.getTableNames().isEmpty()) {
+                leftMessageBuilder.add(format("Tablename: %s", left.getTableNames().toString()));
+            }
+            if (!left.getStageIds().isEmpty()) {
+                leftMessageBuilder.add(format("Stage IDs: %s", left.getStageIds().toString()));
+            }
+
+            if (!right.getTableNames().isEmpty()) {
+                rightMessageBuilder.add(format("Tablename: %s", right.getTableNames().toString()));
+            }
+            if (!right.getStageIds().isEmpty()) {
+                rightMessageBuilder.add(format("Stage IDs: %s", right.getStageIds().toString()));
+            }
+
+            leftMessage = String.join(",", leftMessageBuilder);
+            rightMessage = String.join(",", rightMessageBuilder);
+
+            String warning = format("Try swapping the join ordering for: (%s) %s (%s) in StageId: %s",
```

Comment:
You could also check if stats are available for join inputs and suggest using CBO instead.